ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpImplTreeBuilder.fs

```diff
@@ -1073,12 +1073,19 @@ type FSharpImplTreeBuilder(lexer, document, decls, lifetime, projectedOffset) =
         x.PushRangeForMark(range, mark, ElementType.MATCH_CLAUSE)
 
         x.ProcessPat(pat, true, false)
-        match whenExprOpt with
+        x.PushExpression(expr)
+        x.ProcessWhenExpr(whenExprOpt)
+            
+    member x.ProcessWhenExpr(whenExpr) =
+        match whenExpr with
+        | None -> ()
         | Some whenExpr ->
-            x.PushExpression(expr)
-            x.ProcessExpression(whenExpr)
-        | _ ->
-            x.ProcessExpression(expr)
+            
+        let range = whenExpr.Range
```

Comment:
JFYI there's a helper active pattern `ExprRange` that can be used right in match pattern.